split large jit into several graphs

nimlgen opened the pull request:
gpt-2 timings with CUDA:

new:
ran model in 1.79 ms
total 11.74 ms

old:
ran model in 1.67 ms
total 13.05 ms

geohot commented on tinygrad/jit.py:
```diff
@@ -74,18 +73,29 @@ def __call__(self, *args, **kwargs) -> ReturnType:
-      else:
-        if DEBUG >= 1: print(f"JIT captured {len(self.jit_cache)} kernels with {len(input_rawbuffers)} inputs")
+        # Split JIT cache into batches for faster graph execution. This allows the accelerator to run some batches while subsequent graphs are still being updated.
+        jit_batches: List[List[JitItem]] = []
```
Why are these two loops?

geohot commented on tinygrad/jit.py:
```diff
@@ -22,7 +22,6 @@ def get_input_replace(jit_cache: List[JitItem], input_rawbuffers:List[Buffer]) -
     for i,a in enumerate(ji.rawbufs):
       if a in input_rawbuffers:
         input_replace[(j,i)] = input_rawbuffers.index(a)
-  assert len(set(input_replace.values())) == len(input_rawbuffers), "some input tensors not found"
```
Why was this removed?

nimlgen replied:
This method is called in graphs which contain a part of jit cache which might not have all inputs in their slice. I moved this check to the jit, so the behaviour is the same as it was for a single graph/regular jit.

geohot commented on test/test_jit.py:
```diff
@@ -253,5 +255,31 @@ def foo(x) -> Tensor: return x.sum().realize()
+    for _ in range(5): jf(a, b, c, d).numpy()
+
+    graph_t = Device[Device.DEFAULT].graph.func if isinstance(Device[Device.DEFAULT].graph, functools.partial) else Device[Device.DEFAULT].graph
+    assert len(jf.jit_cache) == 3, "2 graphs and rand op"
```
Whenever you write tests, I want you to think about how many refactors they will survive.

What's going to happen when we switch rand to threefry?

Many of your tests just end up getting deleted because they are fragile.

geohot commented on tinygrad/jit.py:
```diff
@@ -74,16 +73,22 @@ def __call__(self, *args, **kwargs) -> ReturnType:
+        for start in range(0, len(self.jit_cache), batch_size):
+          current_batch = self.jit_cache[start:start+batch_size]
+          try:
+            graphed_jit_cache.append(JitItem(make_graph(current_batch, input_rawbuffers, var_vals), cast(List[Optional[Buffer]], input_rawbuffers)))
```
Err, this won't be good for copies.

geohot commented:
Can you write a spec of what you want this to do?

nimlgen commented:
Yeah that was not ready. Sorry, should have marked as draft.

Basically we want 2 things: 1) split jit into smaller graphs and 2) create graphs even when there are some operations not supported by the graph.  In the current implementation all non-CompiledASTRunner operations are expected to fail the graph creation.

The algo is pretty simple, just iterating through all jit items, combining consecutive items into batches of the maximum allowed size. If a JIT item is a non-CompiledASTRunner, we flush the current batch (which means creating a graph and putting it into the cache) and then add this jit item as is.

This is all implemented in a single pass. I hope it's more or less understandable. I've also added some comments to clarify this concept in the code.

geohot commented:
Much better!
